src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSpark.java

```diff
@@ -0,0 +1,105 @@
+import org.broadinstitute.hellbender.engine.spark.SparkCommandLineProgram;
+import org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSink;
+import org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSource;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.markduplicates.DuplicationMetrics;
+import org.broadinstitute.hellbender.utils.read.markduplicates.OpticalDuplicateFinder;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+
+@CommandLineProgramProperties(
+        summary ="Marks duplicates on spark",
+        oneLineSummary ="Mark Duplicates",
+        programGroup = SparkProgramGroup.class)
+public final class MarkDuplicatesSpark extends SparkCommandLineProgram {
+    private static final long serialVersionUID = 1L;
+
+    //Bases below this quality will not be included in picking the best read from a set of duplicates.
+    private static final int MIN_BASE_QUAL = 15;
+
+    // Used to set an attribute on the GATKRead marking this read as an optical duplicate.
+    private static final String OPTICAL_DUPLICATE_TOTAL_ATTRIBUTE_NAME = "OD";
+
+    @Argument(doc = "uri for the input bam, either a local file path or a gs:// bucket path",
+            shortName = StandardArgumentDefinitions.INPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.INPUT_LONG_NAME,
+            optional = false)
+    protected String bam;
+
+    @Argument(doc = "the output bam", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
+    protected String output;
+
+    @Argument(doc = "File to write duplication metrics to.", optional=true,
+            shortName = "M", fullName = "METRICS_FILE")
+    protected File metricsFile;
+
+    @ArgumentCollection
+    protected OpticalDuplicatesArgumentCollection opticalDuplicatesArgumentCollection = new OpticalDuplicatesArgumentCollection();
+
+    @ArgumentCollection
+    protected IntervalArgumentCollection intervalArgumentCollection = new OptionalIntervalArgumentCollection();
+
+    public static JavaRDD<GATKRead> mark(final JavaRDD<GATKRead> reads, final SAMFileHeader header,
+                                         final OpticalDuplicateFinder opticalDuplicateFinder) {
+
+        JavaRDD<GATKRead> fragments = reads.filter(v1 -> !isNonPrimary(v1));
+        JavaRDD<GATKRead> nonPrimaryReads = reads.filter(v1 -> isNonPrimary(v1));
+        JavaRDD<GATKRead> pairsTransformed = MarkDuplicatesSparkUtils.transformReads(header, opticalDuplicateFinder, fragments);
+
+        JavaRDD<GATKRead> fragmentsTransformed = MarkDuplicatesSparkUtils.transformFragments(header, fragments);
```

Comment:
Can `MarkDuplicatesSparkUtils.transformReads` and `MarkDuplicatesSparkUtils.transformFragments` be handled together?  This would reduce the number of shuffles as well.  One of the methods handles paired ends, while the other handles fragments alone.  Since the paired end key contains the key for the single fragment, would it be possible to first group by `keyForRead` to generate the pairs, but leave in the non-paired reads, then group by fragment, and process the pairs/single fragments accordingly?